### DISCUSSION TITLE
Observing given barycentric coordinates from Earth

### DISCUSSION THREAD

**Discussion by naavis** [skyfielders/python-skyfield#783] [Q&A] [answered]
I have the barycentric coordinates of a thing in the solar system, and I'd like to know where in the sky it is when observed from a specific location on Earth at a specific time. Below is an example script of how I'm trying to do this, but calling `observe` is throwing this error: `'Barycentric' object has no attribute '_observe_from_bcrs'`

What is the recommended way of doing something like this? I am probably misunderstanding or missing something.

```python
from skyfield.api import load, wgs84
from skyfield.positionlib import Barycentric

planets = load('de421.bsp')
earth = planets['earth']
location = earth + wgs84.latlon(0, 0)

ts = load.timescale()

point_in_time = ts.utc(2022, 8, 1)
point_in_space = Barycentric([1.2345, 1.2345, 1.2345])

location.at(point_in_time).observe(point_in_space)
```

**EDIT:** After some more digging through the docs and source code, I'm starting to believe I need to inherit from `VectorFunction` and use that instead of the bare barycentric coordinates. Is that correct?

**Comment by naavis** [comment, marked as answer]
If I understood things right, I got it working by inheriting from `VectorFunction` like this:

```python
from skyfield.api import load, wgs84
from skyfield.vectorlib import VectorFunction

class ThingInSpace(VectorFunction):
    def __init__(self):
        # Value 0 corresponds to the solar system barycenter according to this: https://rhodesmill.org/skyfield/api-vectors.html#skyfield.vectorlib.VectorFunction.center
        self.center = 0

    @property
    def target(self):
        # This is a property to avoid circular references as described here: https://github.com/skyfielders/python-skyfield/blob/master/skyfield/planetarylib.py#L222
        return self

    def _at(self, t):
        # Position vector
        p = [1.2345, 1.2345, 1.2345]
        # Velocity vector
        v = [0.0, 0.0, 0.0]
        return p, v, self.center, "Thing in space"


planets = load('de421.bsp')
earth = planets['earth']
location = earth + wgs84.latlon(0, 0)

ts = load.timescale()

point_in_time = ts.utc(2022, 8, 1)
point_in_space = ThingInSpace()

astrometric = location.at(point_in_time).observe(point_in_space)
print(astrometric.radec())
```

It'd be great if someone could confirm whether this is the way to go.

**Reply by brandon-rhodes** [reply]
I very much dislike the little tangle created by `_observe_from_bcrs()` and `_at()`, which is one reason those routines have underscores in front of them, but I've never yet come up with a good alternative. The `_observe_from_bcrs()` routine exists because star positions are computed very differently from positions of objects in the Solar System; and `_at()` lets the loop that does light-travel-time computation for Solar System objects be more efficient, since otherwise Skyfield would have to build a whole Position object each time through the light-time loop.

But: given that the tangle exists, you've written your class correctly so that Skyfield understands its position, yes! Please follow up with more comments if you run into any trouble, but I think your approach should work.

**Reply by naavis** [reply]
Thank you!
